rust/chains/hyperlane-cosmos/src/providers/mod.rs

```diff
@@ -82,12 +88,95 @@ impl HyperlaneChain for CosmosProvider {
 
 #[async_trait]
 impl HyperlaneProvider for CosmosProvider {
-    async fn get_block_by_hash(&self, _hash: &H256) -> ChainResult<BlockInfo> {
-        todo!() // FIXME
+    async fn get_block_by_hash(&self, hash: &H256) -> ChainResult<BlockInfo> {
+        let tendermint_hash = Hash::from_bytes(Algorithm::Sha256, hash.as_bytes())
+            .expect("block hash should be of correct size");
+
+        // TODO add proper error handling
+        let response = self
+            .rpc_client
+            .block_by_hash(tendermint_hash)
+            .await
+            .map_err(|_| ChainCommunicationError::from_other_str("generic error"))?;
+
+        let received_hash = H256::from_slice(response.block_id.hash.as_bytes());
+
+        if &received_hash != hash {
+            return Err(ChainCommunicationError::from_other_str(
+                "received incorrect block",
```

Comment:
nit: may be nice to surface the hashes in the error, ditto for the tx hash in `get_txn_by_hash`

Reply:
I'll revisit error handling later.